File: docs-examples/example-kotlin/src/test/kotlin/io/micronaut/mqtt/docs/publisher/acknowledge/PublisherAcknowledgeSpec.kt
package io.micronaut.mqtt.docs.publisher.acknowledge;

import io.kotest.assertions.timing.eventually
import io.kotest.matchers.shouldBe
import io.micronaut.mqtt.AbstractMqttKotest
import kotlinx.coroutines.async
import org.opentest4j.AssertionFailedError
import org.reactivestreams.Subscriber
import org.reactivestreams.Subscription
import java.util.concurrent.atomic.AtomicInteger
import kotlin.time.ExperimentalTime
import kotlin.time.seconds

@ExperimentalTime
class PublisherAcknowledgeSpec : AbstractMqttKotest({

    val specName = javaClass.simpleName

    given("Publisher acknowledgement") {
        val ctx = startContext(specName)
        val successCount = AtomicInteger(0)
        val errorCount = AtomicInteger(0)

        `when`("The messages are published") {
            val productClient = ctx.getBean(ProductClient::class.java)
            val publisher = productClient.sendPublisher("publisher body".toByteArray())
            val future = productClient.sendFuture("future body".toByteArray())
            val deferred = async {
                productClient.sendSuspend("suspend body".toByteArray())
            }

            val listener = ctx.getBean(ProductListener::class.java)

            val subscriber = (object : Subscriber<Void> {
                override fun onSubscribe(subscription: Subscription) { }

                override fun onNext(aVoid: Void) {
                    throw UnsupportedOperationException("Should never be called")
                }

                override fun onError(throwable: Throwable) {
                    // if an error occurs
                    errorCount.incrementAndGet()
                }

                override fun onComplete() {
                    // if the publish was acknowledged
                    successCount.incrementAndGet()
                }
            })
            publisher.subscribe(subscriber)
            future.handle { _, t ->
                if (t == null) {
                    successCount.incrementAndGet()
                } else {
                    errorCount.incrementAndGet()
                }
            }
            deferred.invokeOnCompletion {
                if (it == null) {
                    successCount.incrementAndGet()
                } else {
                    errorCount.incrementAndGet()
                }
            }

            then("The messages are published") {
                eventually(10.seconds, AssertionFailedError::class) {
                    errorCount.get() shouldBe 0
                    successCount.get() shouldBe 3
                    listener.messageLengths.size shouldBe 3
                }
            }
        }

        ctx.stop()
    }
})

File: test-suite-python/src/test/python/micronaut/mqtt/docs/publisher/acknowledge/PublisherAcknowledgeSpec.py
from time import sleep
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test
from org.reactivestreams import Subscriber, Subscription

from .ProductClient import ProductClient
from .ProductListener import ProductListener


@Property(name="spec.name", value="PublisherAcknowledgeSpec")
@MicronautTest(environments=["mqtt"])
class PublisherAcknowledgeSpec:
    product_client: Annotated[ProductClient, Inject]
    listener: Annotated[ProductListener, Inject]

    @Test
    def test_publisher_acknowledgement(self):
        success_count = 0
        error_count = 0

        publisher = self.product_client.send_publisher(b"publisher body")
        future = self.product_client.send_future(b"future body")

        class VoidSubscriber(Subscriber):

            def onSubscribe(self, subscription: Subscription) -> None:
                subscription.request(1)

            def onNext(self, value: None) -> None:
                raise RuntimeError("Should never be called")

            def onError(self, throwable: Exception) -> None:
                # if an error occurs
                nonlocal error_count
                error_count += 1

            def onComplete(self) -> None:
                # if the publish was acknowledged
                nonlocal success_count
                success_count += 1

        publisher.subscribe(VoidSubscriber())

        def when_complete(value, throwable):
            nonlocal success_count, error_count
            if throwable is None:
                success_count += 1
            else:
                error_count += 1

        future.whenComplete(when_complete)

        for _ in range(50):
            if success_count == 2 and len(self.listener.message_lengths) == 2:
                break
            sleep(0.1)
        assert error_count == 0
        assert success_count == 2
        assert len(self.listener.message_lengths) == 2
